TFNetworkLayer.py

```diff
@@ -1107,53 +1111,218 @@ def batch_norm(self, data,
       tf.nn.batch_normalization()
       https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/batch_norm.py
     """
+    from operator import mul
+    from functools import reduce
+
     with tf.variable_scope("batch_norm"):
-      if masked_time:
-        x = data.get_placeholder_flattened(keep_dims=True)
-        mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
-      else:
-        x = data.placeholder
-        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
+      # Creating variables
+      fused = fused and not force_sample
+      # Perform batch_norm operations in NHWC format in order to
+      # keep consistency when saving/loading the variables.
+      # Not relevant when fused or masked_time.
+      input_data = data.copy() if fused or masked_time else data.copy_as_batch_spatial_major()
+      ax = input_data.get_axes_from_description(axes)
+      opts = dict(zip(ax, [-1] * len(ax)))
+      variables_shape = input_data.get_bc_shape(opts)
```

Comment:
This does not make sense. Why do you use this function `get_bc_shape`? You defined `opts` in a way that this totally does not make sense to use this function at all. Do you want to get a broadcast (bc) shape? But then your `opts` don't make sense, as you will not get any broadcast shape. Do you just want to get the shape? Then why not directly get the shape?